src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -589,166 +443,247 @@ void GrpcAresRequest::OnAresBackupPollAlarm() {
-    if (!(*it)->already_shutdown) {
+  for (const auto& fd_node : fd_node_list_) {
+    if (!fd_node->already_shutdown) {
       GRPC_ARES_WRAPPER_TRACE_LOG("request: %p shutdown fd: %s", this,
-                                  (*it)->polled_fd->GetName());
-      (*it)->polled_fd->ShutdownLocked(status);
-      (*it)->already_shutdown = true;
+                                  fd_node->polled_fd->GetName());
+      fd_node->polled_fd->ShutdownLocked(status);
+      fd_node->already_shutdown = true;
     }
   }
 }
 
-GrpcAresHostnameRequest::GrpcAresHostnameRequest(
+absl::StatusOr<GrpcAresHostnameRequest*> GrpcAresHostnameRequest::Create(
     absl::string_view name, absl::string_view default_port,
-    EventEngine::Duration timeout,
+    absl::string_view dns_server, EventEngine::Duration timeout,
     std::unique_ptr<GrpcPolledFdFactory> polled_fd_factory,
-    EventEngine* event_engine)
-    : GrpcAresRequest(name, timeout, std::move(polled_fd_factory),
-                      event_engine),
-      default_port_(default_port) {}
-
-GrpcAresHostnameRequest::~GrpcAresHostnameRequest() {
-  // Destruction of on_resolve_ may trigger a chain of destruction that may
-  // require ExecCtx in the current thread.
-  grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
-  grpc_core::ExecCtx exec_ctx;
-  { auto c = std::move(on_resolve_); }
-  GRPC_ARES_WRAPPER_TRACE_LOG("request:%p destructor", this);
-}
-
-void GrpcAresHostnameRequest::Start(
-    absl::AnyInvocable<void(absl::StatusOr<Result>)> on_resolve) {
-  // Holds a ref across this function since OnResolve might be called inline
-  // inside ares_gethostbyname and we might be done in that case. This ref needs
-  // to be taken outside of the lock to prevent UAF during unlock if the
-  // destruction of the ref triggers the destruction of the request itself.
-  auto self = Ref(DEBUG_LOCATION, "Start");
-  grpc_core::MutexLock lock(&mu_);
-  GPR_ASSERT(initialized_);
-  on_resolve_ = std::move(on_resolve);
-  GRPC_ARES_WRAPPER_TRACE_LOG(
-      "request:%p c-ares GrpcAresHostnameRequest::Start name=%s, "
-      "default_port=%s",
-      this, name_.c_str(), default_port_.c_str());
-  // Early out if the target is an ipv4 or ipv6 literal.
-  if (ResolveAsIPLiteralLocked()) {
-    return;
-  }
-  // TODO(yijiem): Early out if the target is localhost and we're on Windows.
-
-  // We add up pending_queries_ here since ares_gethostbyname may directly
-  // invoke the callback inline e.g. if there is any error with the input. The
-  // callback will invoke OnResolve with an error status and may start the
-  // shutdown process too early (before the second ares_gethostbyname) if we
-  // haven't added up here.
-  ++pending_queries_;
+    EventEngine* event_engine) ABSL_NO_THREAD_SAFETY_ANALYSIS {
+  auto* request = new GrpcAresHostnameRequest(
+      name, default_port, timeout, std::move(polled_fd_factory), event_engine);
+  absl::StatusOr<std::string> result = request->ParseNameToResolve();
```

Comment:
The `ParseNameToResolve()` and `ParsePort()` methods seem fairly cumbersome, since their names imply that they just do some parsing, but they actually read and modify the object's state.  Also, the fact that they do that means that we wind up creating and destroying an object that we don't need when the input is invalid.  And we are storing fields like `name_` and `default_port_` as data members when we actually don't actually need them beyond the ctor.

I think this code would be clearer if you just inlined it here.  Consider the following alternative:

```
std::string host;
std::string port_str;
if (!SplitHostPort(name, &host, &port_str)) {
  return absl::InvalidArgumentError(absl::StrCat(
      "cannot parse hostname: ", name));
}
if (port_str.empty()) {
  if (default_port.empty()) {
    return absl::InvalidArgumentError(
        absl::StrFormat("No port in name %s or default_port argument",
                        name));
  }
  port_str = default_port;
}
int port;
if (port_str == "http") {
  port = 80;
} else if (port_str == "https") {
  port = 443;
} else if (!absl::SimpleAtoi(port_str, &port)) {
  return absl::InvalidArgumentError(
      absl::StrCat("Failed to parse port in name: ", name));
}
return new GrpcAresHostnameRequest(
    host, port, timeout, std::move(polled_fd_factory), event_engine);
```

I think this would also allow you to remove the `InitializeAresOptions()` method, because that code can instead be inlined into the base class's ctor.

Similar change in all subclasses.

Reply:
In the new implementation, this part is exactly what you are purposing. We don't need to save `name_` or `default_port_` as data member (though we do save `name` and `port` through QueryArg for constructing the `ResolvedAddress` and for logging purpose). The parsing happens in the `AresResolver::LookupHostname` method:

```
void AresResolver::LookupHostname(
    absl::string_view name, absl::string_view default_port,
    EventEngine::DNSResolver::LookupHostnameCallback callback) {
  absl::string_view host;
  absl::string_view port_s;
  if (!grpc_core::SplitHostPort(name, &host, &port_s)) {
    event_engine_->Run(
        [callback = std::move(callback),
         status = absl::InvalidArgumentError(absl::StrCat(
             "Unparseable name: ", name))]() mutable { callback(status); });
    return;
  }
  GPR_ASSERT(!host.empty());
  if (port_s.empty()) {
    if (default_port.empty()) {
      event_engine_->Run([callback = std::move(callback),
                          status = absl::InvalidArgumentError(absl::StrFormat(
                              "No port in name %s or default_port argument",
                              name.data()))]() mutable { callback(status); });
      return;
    }
    port_s = default_port;
  }
  int port = 0;
  if (port_s == "http") {
    port = 80;
  } else if (port_s == "https") {
    port = 443;
  } else if (!absl::SimpleAtoi(port_s, &port)) {
    event_engine_->Run([callback = std::move(callback),
                        status = absl::InvalidArgumentError(absl::StrCat(
                            "Failed to parse port in name: ",
                            name))]() mutable { callback(status); });
    return;
  }
  // TODO(yijiem): Change this when refactoring code in
  // src/core/lib/address_utils to use EventEngine::ResolvedAddress.
  grpc_resolved_address addr;
  const std::string hostport = grpc_core::JoinHostPort(host, port);
  if (grpc_parse_ipv4_hostport(hostport.c_str(), &addr,
                               false /* log errors */) ||
      grpc_parse_ipv6_hostport(hostport.c_str(), &addr,
                               false /* log errors */)) {
    // Early out if the target is an ipv4 or ipv6 literal.
    std::vector<EventEngine::ResolvedAddress> result;
    result.emplace_back(reinterpret_cast<sockaddr*>(addr.addr), addr.len);
    event_engine_->Run(
        [callback = std::move(callback), result = std::move(result)]() mutable {
          callback(std::move(result));
        });
    return;
  }
  grpc_core::MutexLock lock(&mutex_);
  callback_map_.emplace(++id_, std::move(callback));
  auto* resolver_arg = new HostnameQueryArg(this, id_, name, port);
  if (IsIpv6LoopbackAvailable()) {
    ares_gethostbyname(channel_, std::string(host).c_str(), AF_UNSPEC,
                       &AresResolver::OnHostbynameDoneLocked,
                       static_cast<void*>(resolver_arg));
  } else {
    ares_gethostbyname(channel_, std::string(host).c_str(), AF_INET,
                       &AresResolver::OnHostbynameDoneLocked,
                       static_cast<void*>(resolver_arg));
  }
  CheckSocketsLocked();
  MaybeStartTimerLocked();
}
```